### PR TITLE
fix: handle the user code generation duplication

### PR DESCRIPTION
**Pull request by wood-push-melon**
This pull request tries to resolve the user code generation duplication issue. Currently, the implementation will try a maximum of 3 times to generate a unique user code. When it reaches the maximum number of attempts, an error will be thrown.

### CODE REVIEW

File: handler/rfc8628/auth_handler.go [hunk 1]
```diff
@@ -55,3 +52,49 @@ func (d *DeviceAuthHandler) HandleDeviceEndpointRequest(ctx context.Context, dar
...
+
+			dar.GetSession().SetExpiresAt(fosite.UserCode, time.Now().UTC().Add(d.Config.GetDeviceAndUserCodeLifespan(ctx)).Round(time.Second))
+			err = d.Storage.CreateUserCodeSession(ctx, signature, dar.Sanitize(nil))
+			if err != nil {
+				return "", err
+			}
+
+			return code, nil
+		},
+		retry.Attempts(3),
```
Comment (wood-push-melon) [inline handler/rfc8628/auth_handler.go]: Currently, choose three as the maximum number of attempts.

File: handler/rfc8628/auth_handler.go [hunk 2]
```diff
@@ -55,3 +52,49 @@ func (d *DeviceAuthHandler) HandleDeviceEndpointRequest(ctx context.Context, dar
...
+		retry.Context(ctx),
+		retry.Delay(0),
+		retry.DelayType(retry.CombineDelay(retry.FixedDelay)),
+		retry.LastErrorOnly(true),
+		retry.RetryIf(func(err error) bool {
+			return errors.Is(err, sqlcon.ErrUniqueViolation)
```
Comment (wood-push-melon) [inline handler/rfc8628/auth_handler.go]: I DON'T like this introduction of the error from `sqlcon`.

I think it's better to introduce a new error in `errors.go` and use that here. And the `CreateUserCodeSession` method in Hydra will explicitly throw such an error when a `sqlcon.ErrUniqueViolation` error is thrown by `createSession` method.

**Let me know what you think.**

Comment (nsklikas) [reply handler/rfc8628/auth_handler.go]: I think this sounds better. Even though sqlcon comes from an ory project as well, I would prefer the retry logic to be defined entirely on fosite.

File: handler/rfc8628/auth_handler.go [hunk 3]
```diff
@@ -55,3 +52,49 @@ func (d *DeviceAuthHandler) HandleDeviceEndpointRequest(ctx context.Context, dar
...
+	dar.GetSession().SetExpiresAt(fosite.DeviceCode, time.Now().UTC().Add(d.Config.GetDeviceAndUserCodeLifespan(ctx)))
+	if err = d.Storage.CreateDeviceCodeSession(ctx, signature, dar.Sanitize(nil)); err != nil {
+		return "", errorsx.WithStack(fosite.ErrServerError.WithWrap(err).WithDebug(err.Error()))
+	}
+
+	return code, nil
+}
+
+func (d *DeviceAuthHandler) handleUserCode(ctx context.Context, dar fosite.DeviceRequester) (string, error) {
+	userCode, err := retry.DoWithData(
```
Comment (nsklikas) [inline handler/rfc8628/auth_handler.go]: Can you add a comment here explaining why the retry is needed?

**Review by nsklikas — COMMENTED** [review]
lgtm overall, only minor comments. I think the approach described in your comment sounds better

File: handler/rfc8628/auth_handler.go (hunk 1 above)
Comment (shipperizer) [inline handler/rfc8628/auth_handler.go]: I would drop the library and simply use standard go

```go
i := 3
for {
   if i < 0 {
      break
   }
   do stuff, if err --> continue aka next loop round
   i := i -1

}
```

reason is we don't need delay, if we are getting a clash due to same code already existent we simply try again straight, I know random generation potentially depends on time but even a ns should be sufficient to change the random string

File: handler/rfc8628/auth_handler.go (hunk 2 above)
Comment (shipperizer) [reply handler/rfc8628/auth_handler.go]: I d say we don't care, if u look at `handleDevice` we pass up the error, in this case what I would do is `continue` in the case of a SQLcon error up to x times and then bubble it up, if we error at any other point we bubble it up directly 

we won't need to know it s a SQLcon error but just know that it comes from that storage func call

File: handler/rfc8628/auth_handler.go [hunk 4]
```diff
@@ -55,3 +47,39 @@ func (d *DeviceAuthHandler) HandleDeviceEndpointRequest(ctx context.Context, dar
...
+		return "", errorsx.WithStack(fosite.ErrServerError.WithWrap(err).WithDebug(err.Error()))
+	}
+
+	return code, nil
+}
+
+func (d *DeviceAuthHandler) handleUserCode(ctx context.Context, dar fosite.DeviceRequester) (string, error) {
+	var err error
+	var userCode, signature string
+	maxAttempts := 3
```
Comment (shipperizer) [inline handler/rfc8628/auth_handler.go]: let s chuck it at the top of the file as a constant

File: handler/rfc8628/auth_handler.go [hunk 5]
```diff
@@ -55,3 +47,39 @@ func (d *DeviceAuthHandler) HandleDeviceEndpointRequest(ctx context.Context, dar
...
+		userCode, signature, err = d.Strategy.GenerateUserCode(ctx)
+		if err != nil {
+			return "", err
+		}
+
+		dar.GetSession().SetExpiresAt(fosite.UserCode, time.Now().UTC().Add(d.Config.GetDeviceAndUserCodeLifespan(ctx)).Round(time.Second))
+		err = d.Storage.CreateUserCodeSession(ctx, signature, dar.Sanitize(nil))
+		if err == nil {
+			return userCode, nil
+		}
```
Comment (shipperizer) [inline handler/rfc8628/auth_handler.go]:
```suggestion
		
		if err = d.Storage.CreateUserCodeSession(ctx, signature, dar.Sanitize(nil)); err == nil {
			return userCode, nil
		}
```

File: handler/rfc8628/auth_handler.go [hunk 6]
```diff
@@ -55,3 +47,39 @@ func (d *DeviceAuthHandler) HandleDeviceEndpointRequest(ctx context.Context, dar
...
+	}
+
+	return "", errorsx.WithStack(fosite.ErrServerError.WithWrap(err).WithDebug(err.Error()))
```
Comment (shipperizer) [inline handler/rfc8628/auth_handler.go]:
```suggestion
        errMsg := fmt.Sprintf("exceeded user-code generation max attempts %v : %s", MAX_ATTEMPTS, err.Error())
	return "", errorsx.WithStack(fosite.ErrServerError.WithWrap(err).WithDebug(err.Error()))
```

let's enhance the error with our own message, or add a logging on top, whatever works better for you

**Review by shipperizer — COMMENTED** [review]
looks good, i d say rewrite commit message or squash them into one and then it s good to go 

(comments are minor)

**Review by shipperizer — APPROVED** [review]

**Pull request merged**